openamp/libmetal.defs

```diff
@@ -56,6 +56,20 @@ CSRCS += libmetal/lib/version.c
 
 CFLAGS += -DMETAL_INTERNAL
 
+# Check whether the current toolchain supports 64-bit atomic
+ATOMIC_DETECT_CODE := detect_64_atomic.c
+ATOMIC_DETECT_BIN := detect_64_atomic_bin
+DETECT_ATOMIC_SUPPORT = \
+    echo '\#include <stdatomic.h>' > $(ATOMIC_DETECT_CODE); \
```

Comment:
can we avoid generate the temp file to test 64bit?

Reply:
> can we avoid generate the temp file to test 64bit?

I think this is related to the current toolchain, and there may be no better way. any suggestions?

Reply:
@anchao do you have better idea?

Reply:
GCC/CLANG( all GNU ) toolchains have built-in properties to indicate whether support atomic64:

```
(base) archer@archer:~/code/nuttx/n3/nuttx$ arm-none-eabi-gcc -Wstrict-prototypes -Wno-attributes -Wno-unknown-pragmas -Wno-psabi -fomit-frame-pointer -fno-common -Wall -Wshadow -Wundef -ffunction-sections -fdata-sections -g -mlittle-endian -march=armv7e-m -mtune=cortex-m4 -mfpu=fpv4-sp-d16 -mfloat-abi=hard -mthumb -Wa,-mthumb -Wa,-mimplicit-it=always -isystem /home/archer/code/nuttx/n3/nuttx/include -D__NuttX__ -DNDEBUG -D__KERNEL__  -pipe -DMETAL_INTERNAL -dM -E - < /dev/null | grep SYNC
#define __GCC_HAVE_SYNC_COMPARE_AND_SWAP_1 1
#define __GCC_HAVE_SYNC_COMPARE_AND_SWAP_2 1
#define __GCC_HAVE_SYNC_COMPARE_AND_SWAP_4 1
(base) archer@archer:~/code/nuttx/n3/nuttx$ clang -dM -E - < /dev/null  | grep SYNC
#define __GCC_HAVE_SYNC_COMPARE_AND_SWAP_1 1
#define __GCC_HAVE_SYNC_COMPARE_AND_SWAP_2 1
#define __GCC_HAVE_SYNC_COMPARE_AND_SWAP_4 1
#define __GCC_HAVE_SYNC_COMPARE_AND_SWAP_8 1
(base) archer@archer:~/code/nuttx/n3/nuttx$ gcc -dM -E - < /dev/null  | grep SYNC
#define __GCC_HAVE_SYNC_COMPARE_AND_SWAP_1 1
#define __GCC_HAVE_SYNC_COMPARE_AND_SWAP_2 1
#define __GCC_HAVE_SYNC_COMPARE_AND_SWAP_4 1
#define __GCC_HAVE_SYNC_COMPARE_AND_SWAP_8 1
```
Is it more elegant to check atomic64 use `__GCC_HAVE_SYNC_COMPARE_AND_SWAP_8`?

Reply:
@xuxin930 @xiaoxiang781216 
If `__GCC_HAVE_SYNC_COMPARE_AND_SWAP_8` is more reasonable, please contribute to the openamp community

Reply:
```
diff --git a/openamp/libmetal/lib/io.h b/openamp/libmetal/lib/io.h
index ba416dd505..1de65739ec 100644
--- a/openamp/libmetal/lib/io.h
+++ b/openamp/libmetal/lib/io.h
@@ -30,7 +30,8 @@ extern "C" {
  *  @{
  */
 
-#ifdef __MICROBLAZE__
+#if defined(__MICROBLAZE__) || \
+   (defined(__GNUC__) && !defined(__GCC_HAVE_SYNC_COMPARE_AND_SWAP_8))
 #define NO_ATOMIC_64_SUPPORT
 #endif

```

Reply:
this change needs patch libmetal, the hack in cmake need keep until the next release of OpenAMP.

Reply:
Could we use the hot-fix for this patch? there are a lot of patch files in openamp subdirectory

Reply:
https://github.com/apache/nuttx/pull/12590